File: hms-web/src/app/features/insurance/claims/insurance-claim.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../../environments/environment';
import { InsuranceClaim } from './insurance-claim.model';

export type InsuranceClaimInput = Pick<
  InsuranceClaim,
  'patientId' | 'policyNumber' | 'insurerName' | 'claimType' | 'requestedAmount'
>;

@Injectable({ providedIn: 'root' })
export class InsuranceClaimService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/insurance/claims`;

  list(): Observable<InsuranceClaim[]> {
    return this.http.get<InsuranceClaim[]>(this.baseUrl);
  }

  create(claim: InsuranceClaimInput): Observable<InsuranceClaim> {
    return this.http.post<InsuranceClaim>(this.baseUrl, claim);
  }

  approve(id: number, approvedAmount: number): Observable<InsuranceClaim> {
    return this.http.patch<InsuranceClaim>(`${this.baseUrl}/${id}/approve`, { approvedAmount });
  }

  reject(id: number, reason: string): Observable<InsuranceClaim> {
    return this.http.patch<InsuranceClaim>(`${this.baseUrl}/${id}/reject`, { reason });
  }

  cancel(id: number, reason: string): Observable<InsuranceClaim> {
    return this.http.patch<InsuranceClaim>(`${this.baseUrl}/${id}/cancel`, { reason });
  }
}

File: hms-api/src/main/java/com/pms/insurance/service/InsuranceClaimService.java
package com.pms.insurance.service;

import com.pms.common.EntityNotFoundException;
import com.pms.insurance.dto.InsuranceClaimDto;
import com.pms.insurance.entity.InsuranceClaim;
import com.pms.insurance.entity.InsuranceClaimStatus;
import com.pms.insurance.repository.InsuranceClaimRepository;
import com.pms.registration.entity.Patient;
import com.pms.registration.repository.PatientRepository;
import java.time.Year;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * Replaces IPAction.risePreAuthorization/riseEnhancement and
 * insuranceApprove/insuranceCancel, whose legacy DAO built approval-update
 * HQL via string concatenation - migration doc §4.5, risk R9. Parameterized
 * JPA operations here close that injection surface.
 */
@Service
@Transactional(readOnly = true)
public class InsuranceClaimService {

    private final InsuranceClaimRepository repository;
    private final PatientRepository patientRepository;
    private final AtomicInteger sequence = new AtomicInteger(0);

    public InsuranceClaimService(InsuranceClaimRepository repository, PatientRepository patientRepository) {
        this.repository = repository;
        this.patientRepository = patientRepository;
        this.sequence.set((int) repository.count());
    }

    public List<InsuranceClaimDto> findAll() {
        return repository.findAll().stream().map(this::toDto).toList();
    }

    @Transactional
    public InsuranceClaimDto create(InsuranceClaimDto dto) {
        Patient patient = patientRepository.findById(dto.patientId())
                .orElseThrow(() -> new EntityNotFoundException("Patient not found: " + dto.patientId()));

        InsuranceClaim claim = new InsuranceClaim();
        claim.setClaimNumber(nextClaimNumber());
        claim.setPatient(patient);
        claim.setPolicyNumber(dto.policyNumber());
        claim.setInsurerName(dto.insurerName());
        claim.setClaimType(dto.claimType());
        claim.setRequestedAmount(dto.requestedAmount());
        claim.setStatus(InsuranceClaimStatus.PENDING);

        return toDto(repository.save(claim));
    }

    @Transactional
    public InsuranceClaimDto approve(Long id, double approvedAmount) {
        InsuranceClaim claim = requirePending(id);
        claim.setStatus(InsuranceClaimStatus.APPROVED);
        claim.setApprovedAmount(approvedAmount);
        return toDto(repository.save(claim));
    }

    @Transactional
    public InsuranceClaimDto reject(Long id, String reason) {
        InsuranceClaim claim = requirePending(id);
        claim.setStatus(InsuranceClaimStatus.REJECTED);
        claim.setDecisionReason(reason);
        return toDto(repository.save(claim));
    }

    @Transactional
    public InsuranceClaimDto cancel(Long id, String reason) {
        InsuranceClaim claim = requirePending(id);
        claim.setStatus(InsuranceClaimStatus.CANCELLED);
        claim.setDecisionReason(reason);
        return toDto(repository.save(claim));
    }

    private InsuranceClaim requirePending(Long id) {
        InsuranceClaim claim = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Insurance claim not found: " + id));
        if (claim.getStatus() != InsuranceClaimStatus.PENDING) {
            throw new IllegalArgumentException("Claim " + claim.getClaimNumber() + " is already " + claim.getStatus());
        }
        return claim;
    }

    private String nextClaimNumber() {
        return "INS-" + Year.now().getValue() + "-" + String.format("%05d", sequence.incrementAndGet());
    }

    private InsuranceClaimDto toDto(InsuranceClaim claim) {
        Patient patient = claim.getPatient();
        return new InsuranceClaimDto(
                claim.getId(),
                claim.getClaimNumber(),
                patient.getId(),
                (patient.getFirstName() + " " + (patient.getLastName() != null ? patient.getLastName() : "")).trim(),
                claim.getPolicyNumber(),
                claim.getInsurerName(),
                claim.getClaimType(),
                claim.getRequestedAmount(),
                claim.getApprovedAmount(),
                claim.getStatus(),
                claim.getDecisionReason());
    }
}
